### PR TITLE
16. Modify create event endpoint to accept a list of DrinkActions

### LINKED ISSUE #16: Modify create event endpoint to accept a list of DrinkActions
When the user creates an event, they should have the option to add the drinkActions to that event. So a CreateDrinkActionDto[] (except there is no need for the eventId field on the drinkAction dto. find a nice way of removing that field) field should be on the CreateEventDto. The list can be empty, but use classvalidator to validate the nested object.
Modify the create method of event service to create both the event and the drinkactions ([prisma docs](https://www.prisma.io/docs/orm/prisma-client/queries/relation-queries#nested-writes))

There's no need to modify the update endpoint, the POST/PATCH/DELETE endpoints of the DrinkAction controller can be used to modify the drinkActions of an event once it's created.

depends on #10

### PR DESCRIPTION
**Pull request by murtagh27**

### CODE REVIEW

File: apps/backend/src/drink-action/dto/drink-action.dto.ts [hunk 1]
```diff
@@ -1,8 +1,10 @@
...
+
 import { Drink } from '../../drinks/entities/drink.entity';
 import { Event } from '../../events/entities/event.entity';
 import { DrinkAction } from '../entities/drink-action.entity';
 
-export class DrinkActionDto extends DrinkAction {
+export class DrinkActionDto extends OmitType(DrinkAction, ['eventId']) {
   drink: Drink;
   event: Event;
 }
```
Comment (Tschonti) [inline apps/backend/src/drink-action/dto/drink-action.dto.ts]: You shouldn't edit this file. This DTO represents the type that the GET endpoints of the drink action controller return.

You need to create a new DTO that is very similar to the CreateDrinkActionDto, but with the eventId (you can use OmitType, like here)

File: apps/backend/src/events/dto/create-event.dto.ts [hunk 2]
```diff
@@ -1,5 +1,15 @@
...
 import { Event } from '../entities/event.entity';
 
-export class CreateEventDto extends OmitType(Event, ['id', 'createdAt', 'ownerId']) {}
+export class CreateEventDto extends OmitType(Event, ['id', 'createdAt', 'ownerId']) {
+  @ApiProperty({ type: [CreateDrinkActionDto], required: false })
+  @IsArray()
+  @ValidateNested({ each: true })
+  @Type(() => CreateDrinkActionDto)
+  @IsOptional()
+  drinkActions?: CreateDrinkActionDto[];
```
Comment (Tschonti) [inline apps/backend/src/events/dto/create-event.dto.ts]: I think this works without `@ApiProperty` and `@Type` as well.
Also, use the new DTO here

File: apps/backend/src/events/events.service.ts [hunk 3]
```diff
@@ -36,7 +53,7 @@ export class EventsService {
     if (!event || (event.ownerId !== user.authSchId && !user.isAdmin)) {
       throw new NotFoundException("Event not found or you don't have permission to update it");
     }
-    return await this.prisma.event.update({ where: { id }, data });
+    return await this.prisma.event.update({ where: { id }, data: {} });
```
Comment (Tschonti) [inline apps/backend/src/events/events.service.ts]: with this change, the update method won't change any of the fields on the record, so please revert this change

**Review by Tschonti — CHANGES_REQUESTED** [review]

File: apps/backend/src/events/dto/create-event.dto.ts [hunk 4]
```diff
@@ -1,5 +1,12 @@
...
 
 import { Event } from '../entities/event.entity';
 
-export class CreateEventDto extends OmitType(Event, ['id', 'createdAt', 'ownerId']) {}
+export class CreateEventDto extends OmitType(Event, ['id', 'createdAt', 'ownerId']) {
+  @IsArray()
+  @ValidateNested({ each: true })
+  @IsOptional()
+  drinkActions?: CreateDrinkActionWithoutEventIdDto[];
+}
```
Comment (Tschonti) [inline apps/backend/src/events/dto/create-event.dto.ts]: I was wrong, the `@Type(() => CreateDrinkActionWithoutEventIdDto)` is required, class validator doesn't recognise this field otherwise

File: apps/backend/src/events/events.service.ts [hunk 5]
```diff
@@ -11,12 +12,28 @@ export class EventsService {
...
+          drinkActions: {
+            create:
+              drinkActions?.map((action) => ({
+                ...action,
+                drink: { connect: { id: action.drinkId } },
+              })) || [],
+          },
+        },
+        include: { drinkActions: true },
+      });
```
Comment (Tschonti) [inline apps/backend/src/events/events.service.ts]: If you try this, you'll see that prisma will throw an error that there's an unexpected field (`drinkId`) on the drinkAction object. I'm not sure why this is a problem for prisma, but it works if you remove that field from the action parameter in line 23-26 (you can use [object destructuring](https://byby.dev/ts-object-destructuring))

File: apps/backend/src/events/events.service.ts [hunk 6]
```diff
@@ -11,12 +12,28 @@ export class EventsService {
...
+    } catch (error) {
+      if (error instanceof PrismaClientKnownRequestError && error.code === 'P2025') {
+        throw new BadRequestException('drinkId or eventId does not exist');
+      }
+      throw error;
```
Comment (Tschonti) [inline apps/backend/src/events/events.service.ts]: if you test this, you'll see that the error code is different for foreign key constraint violation.

**Review by Tschonti — CHANGES_REQUESTED** [review]
Getting there, but it doesn't work yet due to some small annoying errors. Make sure to test your solution before submitting the PR

File: apps/backend/src/events/events.service.ts (hunk 6 above)
Comment (murtagh27) [reply apps/backend/src/events/events.service.ts]: i tested it, and maybe i did it wrong, because i still got P2025 when i used an invalid drinkid. when i used P2003 it was internal server error:l

Comment (Tschonti) [reply apps/backend/src/events/events.service.ts]: You're right, P2025 is the correct code. I must've did something wrong last time

**Review by Tschonti — CHANGES_REQUESTED** [review]
Thanks for the fixes, just one last change request: 😄 
The automatic build failed because the drinkActions were connected to the users in another PR, and with that change, you'll have to provide a user in lines 23-26 in `events.service.ts`. If you update your main branch, then merge the main branch into your branch, you'll see the same error locally. Please merge main, connect the user to the new drinkAction (similar to how you connected the drink) and then this will be ready to go.

**Review by Tschonti — APPROVED** [review]
thanks for sticking with this!
